### ISSUE TITLE
X11 backend incurs undefined behavior

### ISSUE THREAD

**Issue by top-sekret** [glfw/glfw#2047]
OS and version: Debian GNU/Linux **bookworm/sid**, kernel **5.15.24-xanmod1-tt**
Release or commit: 97da62a, also some unknown older version used by Minecraft
Error messages: Depends on compilation parameters, either stack smashing detected or plain `SIGSEGV`
Call stack: (built with flags `-ggdb3 -O0 -fstack-protector-strong`, 97da62a)
```
<...unrelated...>
#4  0x00007ff15372fc60 in __stack_chk_fail () at stack_chk_fail.c:24
#5  0x00007ff0ad895a42 in waitForEvent (timeout=0x7ff151ef49e8) at glfw/src/x11_window.c:104
#6  0x00007ff0ad895a86 in waitForVisibilityNotify (window=0x7ff0331f6e80) at glfw/src/x11_window.c:119
#7  0x00007ff0ad89b7cc in _glfwShowWindowX11 (window=0x7ff0331f6e80) at glfw/src/x11_window.c:2397
#8  0x00007ff0ad88777e in glfwCreateWindow (width=854, height=480, title=0x7ff0311feaee "Minecraft* 1.18.1", monitor=0x0, share=0x0)
    at glfw/src/window.c:246
<...unrelated...>
```

There is a fatal bug in the X11 support, which is hard to trigger and prevents GLFW from functioning.

The issue occurs if the program opens a large number of file descriptors before calling GLFW/X11 in any manner. This is the case for example for heavily modded Minecraft (this is how I encountered this bug, as a native crash of the JVM). Although this situation is far from common, it is possible and should be considered.

The problem lies in the legacy `select()` system call used in the `waitForEvent()` private function. The `fd_set` structure has a fixed size, a constant called `FD_SETSIZE`. If a file descriptor has a number equal to or larger than `FD_SETSIZE`, passing it to `FD_SET` causes undefined behavior. As I mentioned, a lot of files were opened before, so the X11 file descriptor could have been large enough. In my case, `FD_SETSIZE` is `1024`, while the X11 file descriptor is `1106`.

Solving this is problematic, because both the alternatives for `select()`, which are `poll()` and `ppoll()`, have some issues. The former has only millisecond timeout granularity (`select` has microsecond granularity), while the latter is Linux specific. I have modified the code to use `ppoll()` and no longer have been able to reproduce the bug with the change implemented.

**Comment by top-sekret** [comment]
Apparently, this bug is already addressed by #2024, which hasn't been merged yet.

**Comment by elmindreda** [comment]
Thank you for the thorough bug report! This should be fixed now with #2024 and the fixes should be included in the next patch release.
